calyx-py/test/arb.py

```diff
@@ -0,0 +1,136 @@
+        # The same issue occurs repeatedly, so I'm just flagging this one.
+        # Whenever the generated futil has {i/j}.out, I actually just want {i/j}.
+        eq_cell.right = cb.const(32, const)
+
+
+def add_emit_from_mem(comp, mem, ans, suffix):
+    """Adds wiring that puts mem{suffix}[j] into ans."""
+    with comp.group(f"emit_from_mem{suffix}") as emit_from_mem:
+        mem.addr0 = CompPort(CompVar("j"), "out")  # AM: want j, not j.out
+        ans.write_en = 1
+        ans.write_data = mem.read_data
+        emit_from_mem.done = ans.done
+
+
+def add_wrap(prog):
+    """Inserts the component `wrap` into the program.
+
+    It has:
+    - two inputs, `i` and `j`
+    - two ref memories, `mem1` and `mem2`
+    - one output, `out`
+
+    For now, assume 0 <= i < 2 and 0 <= j < 4.
+    if i == 0, then out = mem1[j]
+    if i == 1, then out = mem2[j]
+    """
+
+    wrap = prog.component("wrap")
+    wrap.input("i", 32)
+    wrap.input("j", 32)
+
+    mem1 = wrap.mem_d1("mem1", 32, 4, 32, is_ref=True)
+    mem2 = wrap.mem_d1("mem2", 32, 4, 32, is_ref=True)
+    ans = wrap.mem_d1("ans", 32, 1, 32, is_ref=True)
+
+    add_eq(wrap, "i", 0, "eq0", "i_eq_0")
+    add_eq(wrap, "i", 1, "eq1", "i_eq_1")
+    add_emit_from_mem(wrap, mem1, ans, "1")
+    add_emit_from_mem(wrap, mem2, ans, "2")
+
+    # AM, quality of life:
+    # I'd like to generate these `if` statements with the builder.
+    # I tried:
+    #   _: If = cb.if_(
+    #     port=CompPort(CompVar("eq0"), "out"),
+    #     cond=wrap.get_group("i_eq_0"),
+    #     body=Enable("emit_from_mem1"),
+    #   )
+    # but I'm running into trouble with the `port` field, which must be
+    # an ExprBuilder. On digging it looks like that's an GuardExpr in the AST,
+    # but I got stuck at that point.
+    wrap.control = ParComp(
+        [
+            If(
+                port=CompPort(CompVar("eq0"), "out"),
+                cond=CompVar("i_eq_0"),
+                true_branch=Enable("emit_from_mem1"),
+            ),
+            If(
+                port=CompPort(CompVar("eq1"), "out"),
+                cond=CompVar("i_eq_1"),
+                true_branch=Enable("emit_from_mem2"),
+            ),
+        ]
+    )
+
+
+def add_main(prog):
+    """Inserts the component `main` into the program.
+    This will be used to `invoke` the component `wrap`.
+
+    For now, I'd like to pass it memory cells `A` and `B` by reference,
+    along with the inputs i = 1, j = 3.
+    """
+    main = prog.component("main")
+    _ = main.mem_d1("A", 32, 4, 32, is_external=True)
+    _ = main.mem_d1("B", 32, 4, 32, is_external=True)
+    _ = main.mem_d1("out", 32, 1, 32, is_external=True)
+
+    # AM, point of failure (but a stupid hack has worked):
+    # I'd like to add the following to the `cells` section:
+    # together = wrap();
+
+    # The following is clearly a hilarious hack:
+    _ = main.cell("together", CompVar("wrap()"))
```

Comment:
Another thing for @EclecticGriffin potentially!

Reply:
Some relevant advice here: https://github.com/cucapra/calyx/pull/1487#discussion_r1200814552